### ISSUE TITLE
Not all View life cycle events firing for tabs

### ISSUE THREAD

**Issue by mhartington** [driftyco/ionic#2869]
When changing tabs, certain view life-cycle events do not get executed. 

http://codepen.io/mhartington/pen/qEqyBR

If you switch to the about tab, you'll get an alert from the beforeEnter event, but no alert form the beforeLeave event.

**Comment by Ionitron** [comment]
Greetings, @mhartington!

We are now requiring all issues to be submitted through our custom issue form so that we have all of the information necessary to fix them as quickly as possible.

Your progress on this issue has been saved at the following url:

[link to Ionic issue form](http://ionicframework.com/submit-issue/#/?iid=2869)

This issue will be closed if not resubmitted through the form. Please fill out the rest of the form so that we may assist you.

Thank you for allowing me to assist you.

**Comment by jameslegue** [comment]
+1 on this bug - would appreciate a fix or potential work around if available.

**Comment by kepo80** [comment]
+1

**Comment by ag-secret** [comment]
+1
$ionicView unload, beforeLeave, afterLeave and leave don't fire on tab change view.

**Comment by madgrizzle** [comment]
This fix doesn't seem to work for me.  I created a new app using the tabs started and under the 'FriendsCtrl' I added a series of $scope.$on('$ionicView.enter', ....) functions that logs enter, leave, beforeLeave, afterLeave for both $ionicView and $ionicNavView.  Besides using the latest nightly, this is the only change I made to the template (I'll mention I also use Crosswalk).

When I click the friends tab, I get events firing as follows:
$ionicNavView.beforeLeave
$ionicView.enter
$ionicNavView.afterLeave
$ionicNavView.leave

clicking on another tab, I get nothing (no 'leaves').

clicking on a "friend" within the tab, i get
$ionicView.beforeLeave
$ionicView.afterLeave
$ionicView.leave

So.. I wasn't expecting to see any NavView leaves when I enter the and was expecting some sort of leaves when I picked on a different tab, like I did when I picked a friend.  If I don't use the nightly, then I don't get the $ionicNavView.before/after/leave when I click on the Friends tab.  So it appears all the fix does is fires the NavView leaves when you enter???  I tried to figure out the code in the nightly but so far it's beyond my capabilities.

**Comment by jianbo** [comment]
Still having the problem nightly

**Comment by jameslegue** [comment]
Agree - problem still persists and @madgrizzle made a good explanation of what the root cause appears to be.

**Comment by madgrizzle** [comment]
This fix fires the leave events in the leaving controller but still fires the leave events in the arriving controller as well (still that issue with the ViewSwitcher.js change in 9cc61ec)

**Comment by Alkashi** [comment]
Issue not fixed with this update regarding my app.

**Comment by mhartington** [comment]
Seeing this as well

cc @adamdbradley

**Comment by mica16** [comment]
Wow, I was very very anxious about this bug I just figured this morning.
It causes memory leaks since resources were initialized and not closed each time view is generated.
I was using the beforeLeave event. 
Many thanks @mhartington !! 
Will update to rc1 :)

**Comment by mica16** [comment]
Wow, I was very very anxious about this bug I just noticed this morning.
It causes memory leaks since resources were initialized and not closed each time view is generated.
I was using the beforeLeave event. 
I've just updated to rc1, but it doesn't seem to fix it.

Basically I have a view on Tab A.
Its controller declares this: 

```
$scope.$on('$ionicView.beforeLeave', function () {
  alert("life cycle Ok");
});
```
When I simply go back (in the same tab, assuming I was forward in the history), it works, the alert is shown.

However if I DIRECTLY click on tab B, the callback is not called: no alert.

**Comment by mica16** [comment]
Ok, I've found the use case that reproduces the bug. 
Regarding this http://codepen.io/mica16/pen/jEQGvO.  

Working scenario: 
From the first view of the first tab, just click on the "About" tab. 
BeforeEnter and BeforeLeave event are played. 

Buggy scenario: 
From the SECOND view of the first tab (meaning forward in the history), just click on the "About" tab.
Only BeforeEnter is played ! No BeforeLeave.... 

That's the whole of my issue since I need to trigger this leave event from a deep view (meaning forward in the history) in one specific tab.

**Comment by mica16** [comment]
Sorry for all those (my) posts, but I found the real reason: 
my current view was not cached (cache: false), explaining why beforeLeave was not triggered when changing tab. 
Only when you go back or go forward in the same tab, it is triggered.

So I have to deal with $scope.$on('$destroy') in order to make the job ;)

Therefore, it's not a bug.

Thanks !

**Comment by cranesandcaff** [comment]
I have a similar possibly the same issue, All of the beforeEnter, enter, and afterEnter events fire however none of the beforeLeave, leave, or afterLeave events fire. However the $destory events do fire.

**Comment by adamdbradley** [comment]
@mica16 In the demo, the `HomeCtrl` is only assigned to the home view, that happens to be in the tab that's labeled home. The `tabs.facts2` route doesn't have a controller, so it's never creates a beforeLeave event.

**Comment by ioudas210** [comment]
I have this issue in ionic 1.0.0

**Comment by mhartington** [comment]
Could you produce an example of this?

**Comment by pluswave** [comment]
I think I just faced the same issue:

for leaving the whole tabs scope,  the "beforeLeave" event is fired at the scope associated with the tabs, but not a detailed tab.

**Comment by mhartington** [comment]
Just want to stress that everything in the view events is async, so they're really tough to pin point when exactly they happen. So even if they're being reported back as happening after another event, they really aren't.

**Comment by Orthocenter** [comment]
@mhartington I have a similar issue in nested view. Say a.b is abstract view, if I transition from state a.b.c to state a, '$ionic.beforeLeave' will only be fired on a.b state but not a.b.c. Is this a expected behavior?

**Comment by mhartington** [comment]
@Orthocenter try using `$ionicNavView` when dealing with abstract states.

**Comment by cseer90** [comment]
@mhartington: Tried your suggestion but that doesn't work. Basically, I've four tabs. I want to show a loader whenever tab is changed and hide it when the tab is loaded. While `$ionicView.enter` is working finem, `$ionicView.beforeLeave` isn't working at all. Any suggestions? Thanks

**Comment by jsappme** [comment]
try enabling the cache view, it will work. it does not work when cache = false

**Comment by devilankur18** [comment]
+1 Doesn't work when cache = false

**Comment by SeanHub** [comment]
Was this problem every fixed? Its happening for me on 1.3.1

**Comment by Herb-sh** [comment]
I found a work around to this, have a look http://stackoverflow.com/questions/28938408/ionic-view-leave-events-never-called/42905435#42905435.

Hope it helps!
